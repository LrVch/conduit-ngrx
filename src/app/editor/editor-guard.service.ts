import { Injectable } from '@angular/core';
import {
    CanActivate,
    ActivatedRouteSnapshot
} from '@angular/router';
import { Observable, of, throwError } from 'rxjs';
import { Store, select } from '@ngrx/store';
import { AppState } from '@app/reducers';
import { Router } from '@angular/router';
import { tap, take, map, switchMap, catchError, withLatestFrom, filter, mergeMap } from 'rxjs/operators';
import { selectArticle } from './editor.selectors';
import { ArticlesService } from '@app/core';
import { EditorArticleLoadSuccess } from './editor.actions';
import { selectUser } from '@app/auth/auth.selectors';


@Injectable()
export class EditorGuard implements CanActivate {
    constructor(
        private store: Store<AppState>,
        private router: Router,
        private articlesService: ArticlesService
    ) {

    }
    canActivate(
        route: ActivatedRouteSnapshot
    ): Observable<boolean> {

        const slug = route.params['slug'];

        return this.store.pipe(
            take(1),
            switchMap(() => this.articlesService.get(slug)),
            withLatestFrom(this.store.select(selectUser)),
            mergeMap(([article, user]) => article.author.username === user.username
            ? of(new EditorArticleLoadSuccess({ article })) : throwError('wrong user')),
            tap((action: EditorArticleLoadSuccess) => this.store.dispatch(action)),
            map(article => !!article),
            catchError((error) => {
                this.router.navigate(['/']);
                console.error(error);
                return of(false);
            })
        );
    }
}
