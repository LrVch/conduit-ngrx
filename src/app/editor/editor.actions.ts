import { Action } from '@ngrx/store';
import { Article, Errors } from '@app/core';

export enum EditorActionTypes {
  EditorArticleLoadSuccess = '[Editor] Editor Article Load Success',
  EditorArticleLoadFail = '[Editor] Editor Article Load Fail',
  EditorArticleClear = '[Editor] Editor Article Clear',

  EditorArticleSaveRequest = '[Editor] Editor Article Save Request',
  EditorArticleSaveSuccess = '[Editor] Editor Article Save Success',
  EditorArticleSaveFail = '[Editor] Editor Article Save Fail',

  ClearEditorErrors = '[Editor] Clear Editor Errors',
  SetEditorErrors = '[Editor] Set Editor Errors'
}

export class EditorArticleLoadSuccess implements Action {
  readonly type = EditorActionTypes.EditorArticleLoadSuccess;

  constructor(public payload: { article: Article }) {}
}

export class EditorArticleClear implements Action {
  readonly type = EditorActionTypes.EditorArticleClear;
}

export class EditorArticleSaveRequest implements Action {
  readonly type = EditorActionTypes.EditorArticleSaveRequest;

  constructor(public payload: { article: Article }) {}
}

export class EditorArticleSaveSuccess implements Action {
  readonly type = EditorActionTypes.EditorArticleSaveSuccess;

  constructor(public payload: { article: Article }) {}
}

export class EditorArticleSaveFail implements Action {
  readonly type = EditorActionTypes.EditorArticleSaveFail;

  constructor(public payload: { errors: Errors }) {}
}

export class ClearEditorErrors implements Action {
  readonly type = EditorActionTypes.ClearEditorErrors;
}

export class SetEditorErrors implements Action {
  readonly type = EditorActionTypes.SetEditorErrors;

  constructor(public payload: { errors: Errors }) {}
}

export type EditorActions =
  | EditorArticleLoadSuccess
  | EditorArticleClear
  | EditorArticleSaveRequest
  | EditorArticleSaveSuccess
  | EditorArticleSaveFail
  | SetEditorErrors
  | ClearEditorErrors;
