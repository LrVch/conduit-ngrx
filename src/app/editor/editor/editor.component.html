<div class="editor-page page">
  <div class="container ">
    <div class="row">
      <div class="col-md-10 offset-md-1 col-xs-12">

        <app-list-error [errors]="errors$ | async"></app-list-error>

        <div class="mb-4">
          <app-editor-form [article]="article$ | async"
                           [isSubmitting]="isSubmitting$ | async"
                           (onsubmit)="onSubmit($event)"></app-editor-form>
        </div>

      </div>
    </div>
  </div>
</div>