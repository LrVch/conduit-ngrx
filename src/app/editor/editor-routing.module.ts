import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';
import { EditorComponent } from './editor/editor.component';
import { AuthGuard } from '../core';
import { EditorGuard } from './editor-guard.service';
import { CanDeactivateGuard } from '../core/services/can-deactivate.guard';

const routes: Routes = [
  {
    path: '',
    component: EditorComponent,
    canActivate: [AuthGuard],
    canDeactivate: [CanDeactivateGuard]
  },
  {
    path: ':slug',
    component: EditorComponent,
    canActivate: [AuthGuard, EditorGuard],
    canDeactivate: [CanDeactivateGuard]
  }
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class EditorRoutingModule {}
