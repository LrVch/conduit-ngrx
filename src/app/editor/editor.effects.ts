import { Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { ArticlesService } from '../core';
import { EditorActionTypes, EditorArticleSaveRequest, EditorArticleSaveSuccess, EditorArticleSaveFail } from './editor.actions';
import { switchMap, map, catchError, tap } from 'rxjs/operators';
import { of } from 'rxjs';
import { Router } from '@angular/router';
import { AppState } from '../reducers';
import { Store } from '@ngrx/store';
import { ShowMainLoader, HideMainLoader } from '../layout/layout.actions';


@Injectable()
export class EditorEffects {

  constructor(
    private actions$: Actions,
    private articleService: ArticlesService,
    private router: Router,
    private sotre: Store<AppState>
  ) { }

  @Effect()
  saveArticle$ = this.actions$.pipe(
    ofType<EditorArticleSaveRequest>(EditorActionTypes.EditorArticleSaveRequest),
    switchMap(action => this.articleService.save(action.payload.article).pipe(
      map(article => new EditorArticleSaveSuccess({ article })),
      catchError(errors => {
        console.error(errors);
        return of(new EditorArticleSaveFail({ errors }));
      })
    )));

  @Effect({dispatch: false})
  editorArticleSaveSuccess$ = this.actions$.pipe(
    ofType<EditorArticleSaveSuccess>(EditorActionTypes.EditorArticleSaveSuccess),
    map(action => action.payload.article),
    tap(article =>  this.router.navigateByUrl('/article/' + article.slug))
  );

  @Effect({dispatch: false})
  editorArticleSaveRequest$ = this.actions$.pipe(
    ofType<EditorArticleSaveRequest>(EditorActionTypes.EditorArticleSaveRequest),
    tap(_ => this.sotre.dispatch(new ShowMainLoader())),
  );

  @Effect({dispatch: false})
  editorArticleSaveFail$ = this.actions$.pipe(
    ofType<EditorArticleSaveFail>(EditorActionTypes.EditorArticleSaveFail),
    tap(_ => this.sotre.dispatch(new HideMainLoader()))
  );
}
