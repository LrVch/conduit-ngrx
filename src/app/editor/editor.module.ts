import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { EditorComponent } from './editor/editor.component';
import { SharedModule } from '../shared';
import { EditorRoutingModule } from './editor-routing.module';
import { EditorFormComponent } from './editor-form/editor-form.component';
import { StoreModule } from '@ngrx/store';
import * as fromEditor from './editor.reducer';
import { EffectsModule } from '@ngrx/effects';
import { EditorEffects } from './editor.effects';
import { EditorGuard } from './editor-guard.service';
import { TranslateModule, TranslateLoader } from '@ngx-translate/core';
import { TranslateHttpLoader } from '@ngx-translate/http-loader';
import { HttpClient } from '@angular/common/http';
import { environment } from '@env/environment';

@NgModule({
  imports: [
    CommonModule,
    SharedModule,
    EditorRoutingModule,
    StoreModule.forFeature('editor', fromEditor.editorReducer),
    EffectsModule.forFeature([EditorEffects]),
    TranslateModule.forChild({
      loader: {
        provide: TranslateLoader,
        useFactory: HttpLoaderFactory,
        deps: [HttpClient]
      }
    }),
  ],
  declarations: [EditorComponent, EditorFormComponent],
  providers: [EditorGuard]
})
export class EditorModule { }

export function HttpLoaderFactory(http: HttpClient) {
  return new TranslateHttpLoader(
    http,
    `${environment.i18nPrefix}/assets/i18n/`,
    '.json'
  );
}
