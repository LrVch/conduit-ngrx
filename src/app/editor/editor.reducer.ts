import { Action } from '@ngrx/store';
import { Article, Errors } from '../core';
import { EditorActionTypes, EditorActions } from './editor.actions';


export interface EditorState {
  article: Article;
  savingArticle: boolean;
  errors: Errors;
}

export const initialState: EditorState = {
  article: null,
  savingArticle: false,
  errors: {
    errors: {}
  }
};

export function editorReducer(state = initialState, action: EditorActions): EditorState {
  switch (action.type) {

    case EditorActionTypes.EditorArticleLoadSuccess:
    case EditorActionTypes.EditorArticleSaveSuccess:
      return {
        ...state,
        article: action.payload.article,
        savingArticle: false,
        errors: {
          errors: {}
        }
      };

    case EditorActionTypes.EditorArticleClear:
      return {
        ...state,
        article: null
      };
    case EditorActionTypes.EditorArticleSaveRequest:
      return {
        ...state,
        savingArticle: true,
        errors: {
          errors: {}
        }
      };
    case EditorActionTypes.EditorArticleSaveFail:
      return {
        ...state,
        savingArticle: false,
        errors: { ...action.payload.errors }
      };
    case EditorActionTypes.SetEditorErrors:
      return {
        ...state,
        errors: { ...action.payload.errors }
      };
    case EditorActionTypes.ClearEditorErrors:
      return {
        ...state,
        errors: {
          errors: {}
        }
      };

    default:
      return state;
  }
}
