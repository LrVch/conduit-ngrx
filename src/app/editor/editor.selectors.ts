import { createSelector, createFeatureSelector } from '@ngrx/store';
import { AppState } from '../reducers';
import { EditorState } from './editor.reducer';

export const selectEditorState = createFeatureSelector<AppState, EditorState>('editor');

export const selectArticle = createSelector(
  selectEditorState,
  (editorState: EditorState) => editorState.article
);

export const selectArticleSaving = createSelector(
  selectEditorState,
  (editorState: EditorState) => editorState.savingArticle
);

export const selectEditorErrors = createSelector(
  selectEditorState,
  (editorState: EditorState) => editorState.errors
);
