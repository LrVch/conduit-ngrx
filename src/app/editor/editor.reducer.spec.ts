import * as fromEditor from './editor.reducer';
import {
  EditorArticleLoadSuccess,
  EditorArticleSaveSuccess,
  EditorArticleClear,
  EditorArticleSaveRequest,
  EditorArticleSaveFail,
  ClearEditorErrors
} from './editor.actions';
import { getArticle, getSomeErrors } from '../lib/testing';

describe('Editor Reducer', () => {

  describe('State changes', () => {
    const article = getArticle();
    const errors = getSomeErrors();
    const initialEditorState: fromEditor.EditorState = fromEditor.initialState;

    it('should have an initial state', () => {
      const state = fromEditor.editorReducer(initialEditorState, { type: '@@init' } as any);

      expect(state).toBe(initialEditorState);
    });

    it('should set article on "EditorArticleLoadSuccess" or "EditorArticleSaveSuccess"', () => {
      const init = { type: '@@init' } as any;
      const action1 = new EditorArticleLoadSuccess({ article });
      const action2 = new EditorArticleSaveSuccess({ article });

      const state1 = [init, action1].reduce(fromEditor.editorReducer, initialEditorState);
      const state2 = [init, action2].reduce(fromEditor.editorReducer, initialEditorState);

      expect(state1).toMatchSnapshot();
      expect(state2).toMatchSnapshot();
    });

    it('should remove article on "EditorArticleClear"', () => {
      const init = { type: '@@init' } as any;
      const action = new EditorArticleClear();

      const state = [init, action].reduce(fromEditor.editorReducer, initialEditorState);

      expect(state).toMatchSnapshot();
    });

    it('should set savingArticle as true on "EditorArticleSaveRequest"', () => {
      const init = { type: '@@init' } as any;
      const action = new EditorArticleSaveRequest({ article });

      const state = [init, action].reduce(fromEditor.editorReducer, initialEditorState);

      expect(state).toMatchSnapshot();
    });

    it('should set savingArticle as false on "EditorArticleSaveFail"', () => {
      const init = { type: '@@init' } as any;
      const action = new EditorArticleSaveFail({ article });

      const state = [init, action].reduce(fromEditor.editorReducer, initialEditorState);

      expect(state).toMatchSnapshot();
    });

    it('should set errors on "SetEditorErrors"', () => {
      const init = { type: '@@init' } as any;
      const action = new EditorArticleSaveFail({ errors });

      const state = [init, action].reduce(fromEditor.editorReducer, initialEditorState);

      expect(state).toMatchSnapshot();
    });

    it('should remove errors on "ClearEditorErrors"', () => {
      const init = { type: '@@init' } as any;
      const action = new ClearEditorErrors();

      const state = [init, action].reduce(fromEditor.editorReducer, initialEditorState);

      expect(state).toMatchSnapshot();
    });
  });
});
