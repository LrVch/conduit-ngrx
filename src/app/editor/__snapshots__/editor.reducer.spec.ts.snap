// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor Reducer State changes should remove article on "EditorArticleClear" 1`] = `
Object {
  "article": null,
  "errors": Object {
    "errors": Object {},
  },
  "savingArticle": false,
}
`;

exports[`Editor Reducer State changes should remove errors on "ClearEditorErrors" 1`] = `
Object {
  "article": null,
  "errors": Object {
    "errors": Object {},
  },
  "savingArticle": false,
}
`;

exports[`Editor Reducer State changes should set article on "EditorArticleLoadSuccess" or "EditorArticleSaveSuccess" 1`] = `
Object {
  "article": Object {
    "author": Object {
      "bio": "bio",
      "following": false,
      "image": "image",
      "isFollowLoading": false,
      "username": "username",
    },
    "body": "body",
    "createdAt": "1544505543896",
    "description": "desc",
    "favorited": false,
    "favoritesCount": 0,
    "favoriting": false,
    "slug": "slug",
    "tagList": Array [
      "list",
    ],
    "title": "title",
    "updatedAt": "1544505523896",
  },
  "errors": Object {
    "errors": Object {},
  },
  "savingArticle": false,
}
`;

exports[`Editor Reducer State changes should set article on "EditorArticleLoadSuccess" or "EditorArticleSaveSuccess" 2`] = `
Object {
  "article": Object {
    "author": Object {
      "bio": "bio",
      "following": false,
      "image": "image",
      "isFollowLoading": false,
      "username": "username",
    },
    "body": "body",
    "createdAt": "1544505543896",
    "description": "desc",
    "favorited": false,
    "favoritesCount": 0,
    "favoriting": false,
    "slug": "slug",
    "tagList": Array [
      "list",
    ],
    "title": "title",
    "updatedAt": "1544505523896",
  },
  "errors": Object {
    "errors": Object {},
  },
  "savingArticle": false,
}
`;

exports[`Editor Reducer State changes should set errors on "SetEditorErrors" 1`] = `
Object {
  "article": null,
  "errors": Object {
    "errors": Object {
      "error": Array [
        "some error",
      ],
    },
  },
  "savingArticle": false,
}
`;

exports[`Editor Reducer State changes should set savingArticle as false on "EditorArticleSaveFail" 1`] = `
Object {
  "article": null,
  "errors": Object {},
  "savingArticle": false,
}
`;

exports[`Editor Reducer State changes should set savingArticle as true on "EditorArticleSaveRequest" 1`] = `
Object {
  "article": null,
  "errors": Object {
    "errors": Object {},
  },
  "savingArticle": true,
}
`;
